service/src/main/java/bio/terra/tanagra/db/AllArtifactsDao.java

```diff
@@ -0,0 +1,195 @@
+import bio.terra.tanagra.service.artifact.model.AnnotationValue;
+import bio.terra.tanagra.service.artifact.model.Cohort;
+import bio.terra.tanagra.service.artifact.model.CohortRevision;
+import bio.terra.tanagra.service.artifact.model.Review;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.apache.commons.lang3.tuple.Pair;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.jdbc.core.namedparam.MapSqlParameterSource;
+import org.springframework.stereotype.Component;
+
+// A Dao for all artifacts: this is needed to avoid circular dependency errors
+@Component
+public class AllArtifactsDao {
+  private final CohortDao cohortDao;
+  private final ReviewDao reviewDao;
+  private final AnnotationDao annotationDao;
+
+  @Autowired
+  public AllArtifactsDao(CohortDao cohortDao, ReviewDao reviewDao, AnnotationDao annotationDao) {
+    this.cohortDao = cohortDao;
+    this.reviewDao = reviewDao;
+    this.annotationDao = annotationDao;
+  }
+
+  /**
+   * Clones a cohort and all related artifacts by copying DB rows. Does not include validations.
+   * Returns the id of the new cohort
+   */
+  @WriteTransaction
+  public String cloneCohort(
+      String cohortId,
+      String userEmail,
+      String destinationStudyId,
+      String displayName,
+      String description) {
+    // Cohort
+    Cohort ogCohort = cohortDao.getCohort(cohortId);
+    String clonedCohortId = ServiceUtils.newArtifactId();
+
+    cohortDao.insertCohortRow(
+        destinationStudyId,
+        clonedCohortId,
+        ogCohort.getUnderlay(),
+        userEmail,
+        displayName,
+        description);
+
+    // Reviews
+    List<Review> ogReviews = reviewDao.getAllReviews(cohortId, 0, Integer.MAX_VALUE);
+    Map<String, String> ogRevisionIdReviewIdMap = new HashMap<>(); // revisionId -> reviewId
+    Map<Integer, String> ogRevisionVerReviewIdMap = new HashMap<>(); // revisionVersion -> reviewId
+    Map<String, String> reviewIdMap = new HashMap<>(); // originalId -> clonedId
+    List<MapSqlParameterSource> reviewParamSets = new ArrayList<>();
+    List<MapSqlParameterSource> primaryEntityParamSets = new ArrayList<>();
+
+    ogReviews.forEach(
+        review -> {
+          String ogReviewId = review.getId();
+          String clonedId = ServiceUtils.newArtifactId();
+          ogRevisionIdReviewIdMap.put(review.getRevision().getId(), ogReviewId);
+          ogRevisionVerReviewIdMap.put(review.getRevision().getVersion(), ogReviewId);
+          reviewIdMap.put(ogReviewId, clonedId);
+
+          reviewParamSets.add(
+              reviewDao.buildReviewParam(
+                  clonedCohortId,
+                  clonedId,
+                  review.getSize(),
+                  review.getDisplayName(),
+                  review.getDescription(),
+                  userEmail));
+
+          Map<Literal, Integer> primaryEntityIdsToStableIndex =
```

Comment:
Here you're reading all the primary entity ids for all reviews for this cohort into memory. Can you copy the rows in SQL instead with an `INSERT ... SELECT` statement to avoid this? You could try joining on the revision version number to get the new review id, or do it one review at a time.

Reply:
done